[SR-88] Fixing and reapplying stdlib mirror API changes

austinzheng opened the pull request:
This PR contains both a commit reverting the revert to the mirrors commit (original pull request [here](https://github.com/apple/swift/pull/838)), as well as another commit containing fixes and changes. I took the opportunity to remove some additional code.

I'm working on running the test suite (unit and validation tests) on both OS X and Ubuntu, and will comment when they are done.

I do have two comments:

* I'm a little surprised that the way generics work changed so drastically in the past few days that the code I had originally written stopped working. If this wasn't intentional, there might be a regression somewhere. I'll have to look through the commit history and try to pinpoint exactly what changed.

* The original mirror for sets and collections cached a 'last accessed' index value and used that to speed up subsequent accesses if those accesses involved 'larger' indices. That optimization is gone now that all the collections in the stdlib use the same helper collection type to pass their children into a mirror initializer. I wanted to point that out in case it might matter.

Let me know if there are any questions. Thanks again!

austinzheng commented:
I forgot to mention that I'll squash/rebase once I verify tests are working and get the go-ahead.

austinzheng commented:
Tests pass on both OS X and Ubuntu.

dabrahams commented:
Did you intend to squash these?

austinzheng commented:
Yeah, eventually. Wanted to give people the chance to look at the changes first.

Austin

> On Jan 23, 2016, at 8:55 AM, Dave Abrahams <notifications@github.com> wrote:
> 
> Did you intend to squash these?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1058#issuecomment-174200026>.
>

dabrahams commented:
on Fri Jan 22 2016, Austin Zheng <notifications-AT-github.com> wrote:

> * The original mirror for sets and collections cached a 'last
> accessed' index value and used that to speed up subsequent accesses if
> those accesses involved 'larger' indices. That optimization is gone
> now that all the collections in the stdlib use the same helper
> collection type to pass their children into a mirror initializer. I
> wanted to point that out in case it might matter.

@egranata, are you still relying on this optimization?  

-- 
-Dave

dabrahams commented:
on Fri Jan 22 2016, Austin Zheng <notifications-AT-github.com> wrote:

> * I'm a little surprised that the way generics work changed so
> drastically in the past few days that the code I had originally
> written stopped working. If this wasn't intentional, there might be a
> regression somewhere. I'll have to look through the commit history and
> try to pinpoint exactly what changed.

I must be missing something.  I don't see anything in your fixes that
look like they point to a change in the way generics work.  Can you
explain?

-- 
-Dave

austinzheng commented:
Sure. What I meant to say was that the code in my original PR was building on my machine because it was based on a commit that was about 50 commits (a few day's worth) behind where `master` actually was at that time. When I updated my local master to match upstream's master, and then rebased my PR branch onto the updated master, only then did it fail to build correctly for me. Because of this I concluded that at least one of those ~50 changes between the two `master` commits contained something that made my original PR code stop working. Since the error in question was a type error involving generics, I imagine one of three things happened:

* There was an incorrect use of generics in the code in my PR branch that wasn't being caught before, but a compiler bug was fixed and that incorrect use started to be caught.
* One of those 50 commits introduced a regression, and my correct use of generics was incorrectly flagged as being an error.
* One of those 50 commits intentionally changed how generics worked, and my previously-valid code now became invalid.

Let me know if that was any clearer. Thanks!

Austin

> On Jan 23, 2016, at 10:37 AM, Dave Abrahams <notifications@github.com> wrote:
> 
> 
> on Fri Jan 22 2016, Austin Zheng <notifications-AT-github.com> wrote:
> 
> > * I'm a little surprised that the way generics work changed so
> > drastically in the past few days that the code I had originally
> > written stopped working. If this wasn't intentional, there might be a
> > regression somewhere. I'll have to look through the commit history and
> > try to pinpoint exactly what changed.
> 
> I must be missing something. I don't see anything in your fixes that
> look like they point to a change in the way generics work. Can you
> explain?
> 
> -- 
> -Dave
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1058#issuecomment-174210993>.
>

dabrahams commented:
It’s a little clearer, but letting us look at the un-squashed commits is only useful for us in tracking that down if you separate the changes that actually fixed breakage from the other stuff.  And my point is, looking at ed69418, I can’t tell what made the difference.

austinzheng commented:
Okay, that makes sense. I'll squash the commits this afternoon.

The breakage-fixing 'change' was changing the Mirrors to not use the helper collections that were deleted in that commit. Sorry for not making that clearer.

Sent from my iPhone

> On Jan 23, 2016, at 1:34 PM, Dave Abrahams <notifications@github.com> wrote:
> 
> It’s a little clearer, but letting us look at the un-squashed commits is only useful for us in tracking that down if you separate the changes that actually fixed breakage from the other stuff. And my point is, looking at ed69418, I can’t tell what made the difference.
> 
> —
> Reply to this email directly or view it on GitHub.
>

austinzheng commented:
I just wanted to know if there's anything I can do to help push this forward. If not, I can ask again in a few days.

gribozavr commented:
@austinzheng Could you resolve conflicts with the master branch?

austinzheng commented:
I'll do that right now; however, I think the points @dabrahams still need to be addressed. Let me know if you have questions.

> On Jan 25, 2016, at 8:26 PM, Dmitri Gribenko <notifications@github.com> wrote:
> 
> @austinzheng <https://github.com/austinzheng> Could you resolve conflicts with the master branch?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/apple/swift/pull/1058#issuecomment-174816496>.
>

gribozavr commented:
Yes, we still need to decide what to do with that optimization in Dictionary's mirror.  I know it is important for the debugger, I want to read the new code and consider how it would perform.

austinzheng commented:
I squashed and rebased onto latest master. I'm running tests right now just to verify nothing is broken. Once you've decided what you want to do with Dictionary's mirror, let me know.

Also, as I mentioned in one of the line comments(?), the collections now use the unlabeledChildren: constructor for `Mirror`, which means `dump()`ing them no longer displays the `[0]`, `[1]`, etc labels. If you don't like that I can bring back the old behavior.

austinzheng commented:
As of now, all tests on this branch pass on both OS X and Ubuntu.

gribozavr commented on stdlib/public/SDK/AppKit/AppKit.swift:
```diff
@@ -13,94 +13,39 @@
+      return .View(NSImage())
+    } else {
+      _NSViewQuickLookState.views.insert(self)
+      let result : PlaygroundQuickLook
```
We don't put a space before the colon in variable declarations.

gribozavr commented on stdlib/public/SDK/CoreGraphics/CoreGraphics.swift:
```diff
@@ -39,6 +39,22 @@ public extension CGPoint {
+}
+
+extension CGPoint : CustomDebugStringConvertible {
+  public var debugDescription : String {
```
Extra space before the colon.

gribozavr commented on stdlib/public/SDK/CoreGraphics/CoreGraphics.swift:
```diff
@@ -68,6 +84,22 @@ public extension CGSize {
+}
+
+extension CGSize : CustomDebugStringConvertible {
+  public var debugDescription : String {
```
Extra space before the colon.

gribozavr commented on stdlib/public/SDK/CoreGraphics/CoreGraphics.swift:
```diff
@@ -357,6 +389,22 @@ public extension CGRect {
+}
+
+extension CGRect : CustomDebugStringConvertible {
+  public var debugDescription : String {
```
Extra space before the colon.

gribozavr commented on stdlib/public/SDK/Foundation/Foundation.swift:
```diff
@@ -1388,3 +1392,62 @@ extension NSURL : _FileReferenceLiteralConvertible {
+
+extension NSSet : CustomReflectable {
+  public func customMirror() -> Mirror {
+    return Mirror(reflecting: _convertNSSetToSet(self) as Set<NSObject>)
```
Why did you have to call `_convertNSSetToSet` explicitly?  This is an internal entry point that shouldn't be called directly.

`self as Set<NSObject>` should work.

austinzheng replied:
Sounds good. I was being conservative and trying to be as close to the original implementation as possible.

gribozavr replied:
Yes, but the old code wasn't necessarily the best or even correct :)

austinzheng replied:
I guess sometimes bad code is just bad code and not super special abstraction-breaking compiler magic :).

gribozavr commented on stdlib/public/SDK/Foundation/Foundation.swift:
```diff
@@ -1388,3 +1392,62 @@ extension NSURL : _FileReferenceLiteralConvertible {
+
+extension NSDictionary : CustomReflectable {
+  public func customMirror() -> Mirror {
+    return Mirror(reflecting: _convertNSDictionaryToDictionary(self) as [NSObject : AnyObject])
```
Same here.  `self as [NSObject : AnyObject]` should work.  No need to call internal entry points.

gribozavr commented on stdlib/public/SDK/SpriteKit/SpriteKitQuickLooks.swift.gyb:
```diff
@@ -0,0 +1,39 @@
+
+%import gyb
+
+% for Self in ['SKShapeNode','SKSpriteNode','SKTextureAtlas','SKTexture']:
```
Please add spaces after commas per Python coding style.

gribozavr commented on stdlib/public/core/Mirrors.swift.gyb:
```diff
@@ -39,11 +39,19 @@ internal func _toString<T>(x: T) -> String {
 
-% for Type in Types:
-extension ${Type[0]} : _Reflectable {
+%import gyb
```
Is `import gyb` really needed?  I don't see what would depend on it.

gribozavr commented on stdlib/public/core/Reflection.swift:
```diff
@@ -187,22 +188,38 @@ func _dumpWithMirror<TargetStream : OutputStreamType>(
+  // This takes the place of the old mirror API's 'summary' property
+  _dumpPrint(object, mirror, &targetStream)
+
+  let id : ObjectIdentifier?
```
Extra space before the colon.

gribozavr commented on stdlib/public/core/Reflection.swift:
```diff
@@ -219,38 +236,62 @@ func _dumpWithMirror<TargetStream : OutputStreamType>(
-// -- _MirrorType implementations for basic data types
+/// Dump information about an object's superclass, given a mirror reflecting
+/// that superclass.
+func _dumpSuperclass<TargetStream : OutputStreamType>(
```
Please add `internal` explicitly.

gribozavr commented on stdlib/public/core/Reflection.swift:
```diff
@@ -167,19 +167,20 @@ public func dump<T>(x: T, name: String? = nil, indent: Int = 0,
-func _dumpWithMirror<TargetStream : OutputStreamType>(
-    mirror: _MirrorType, _ name: String?, _ indent: Int, _ maxDepth: Int,
+/// Dump an object's contents. User code should use dump().
+func _dumpObject<TargetStream : OutputStreamType>(
```
Please add an explicit `internal`.

gribozavr commented on stdlib/public/core/UnsafePointer.swift.gyb:
```diff
@@ -423,34 +417,25 @@ extension ${Self} : CustomDebugStringConvertible {
-    let ptrValue = _getPointerValue()
-    if ptrValue == 0 { return "\(selfType)(nil)" }
-    return "\(selfType)(0x\(_uint64ToString(ptrValue, radix:16, uppercase:true)))"
+    let ptrValue = UInt64(Int(Builtin.ptrtoint_Word(_rawValue)))
```
The old code was wrong for large pointer values (for `> Int64.max` it would crash), let's fix it:

`UInt64(bitPattern: Int(...))`

gribozavr commented on stdlib/public/core/UnsafePointer.swift.gyb:
```diff
@@ -423,34 +417,25 @@ extension ${Self} : CustomDebugStringConvertible {
-    }
+extension ${Self} : CustomReflectable {
+  public func customMirror() -> Mirror {
+    let ptrValue = UInt64(Int(Builtin.ptrtoint_Word(_rawValue)))
```
Same as below, `UInt64(bitPattern: Int(...))`.

gribozavr commented:
@austinzheng Thank you for continuing to push this forward!  I have left a few stylistic and correctness comments.  We also discussed the indexing optimization in Set and Dictionary, and we understood that we don't need it now.  So could you fix the few issues I pointed out?  We are all clear to merge then!

austinzheng commented:
Thanks for the code review, much appreciated! I'll fix the issues you
pointed out tonight, and let you know when everything is ready.

Austin

On Tue, Jan 26, 2016 at 3:12 PM, Dmitri Gribenko <notifications@github.com>
wrote:

> @austinzheng <https://github.com/austinzheng> Thank you for continuing to
> push this forward! I have left a few stylistic and correctness comments. We
> also discussed the indexing optimization in Set and Dictionary, and we
> understood that we don't need it now. So could you fix the few issues I
> pointed out? We are all clear to merge then!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift/pull/1058#issuecomment-175287534>.
>

austinzheng commented:
I made the changes, squashed, and rebased onto latest master. I'm running tests right now on OS X and Ubuntu; probably a good idea to wait until those are done.

I added a few unit tests to ensure that pointers created using bit patterns of values > `Int64.max` don't cause a crash.

gribozavr commented:
@swift-ci Please test Linux platform

austinzheng commented:
There was a typo in my new test case. I fixed it and pushed again. I'll let you know when the tests complete successfully.

gribozavr commented:
@swift-ci Please test Linux platform

austinzheng commented:
My unit test needs some work. Just a few minutes...

austinzheng commented:
All tests now pass on OS X. Linux pending...

austinzheng commented:
Unfortunately I have to head out now, but if you trigger the CI to run on Linux it should almost certainly pass. Let me know if you have any questions.

shahmishal commented:
@swift-ci Please test Linux platform

shahmishal commented:
@swift-ci Please test OS X platform

gribozavr commented:
@swift-ci Please test Linux platform

gribozavr commented:
@austinzheng This is amazing work, thank you!

@swift-ci :+1:

dabrahams commented:
Yes, @austinzheng, thank you again for taking this on!

austinzheng commented:
I'm glad it's finally in! Fingers crossed that it won't break anything :grimacing:.

atrick commented:
It looks like the reflection tests are breaking in the asserts build. You can revert my revert commit below and provide a fix on top of that.

commit 182bb7f812bbb0c103dab80a3b0f6374bbc3593a (HEAD -> master, public/master, public/HEAD)
Author: Andrew Trick <atrick@apple.com>
Date:   Wed Jan 27 10:43:08 2016

    Revert "Merge pull request #1058 from austinzheng/az-port-mirror"
    
    This pull request broke the following tests on several build configurations
    (eg --preset=buildbot,tools=RA,stdlib=DA)
    
        1_stdlib/Reflection.swift
        1_stdlib/ReflectionHashing.swift
        1_stdlib/UnsafePointer.swift.gyb
    
    This reverts commit c223a3bf06e3d4173f0d6cf0ad21c41d344ca1e0, reversing
    changes made to 5c2bb09b092ffd23e2afc8e31a0a7f2f93974aa1.

atrick commented:
The key to reproducing the failures is probably --swift-stdlib-build-type=Debug --swift-stdlib-enable-assertions=true

austinzheng commented:
Thanks for bringing this to my attention; I'll take a look tonight. Are
those arguments to build-script?

Are there any other configs I can test against before submitting a fix?

Austin

On Wed, Jan 27, 2016 at 11:02 AM, atrick <notifications@github.com> wrote:

> The key to reproducing the failures is probably
> --swift-stdlib-build-type=Debug --swift-stdlib-enable-assertions=true
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift/pull/1058#issuecomment-175795126>.
>

atrick commented:
There were two kinds of failures that persisted after this commit:

- stdlib asserts build

Both the release and debug stdlib build with asserts have the same failures:

build-script --preset=buildbot,tools=RA,stdlib=DA 
build-script --preset=buildbot,tools=RA,stdlib=RDA

I'm guessing the relevant build-script option is --swift-stdlib-enable-assertions=true

    1_stdlib/Reflection.swift
    1_stdlib/ReflectionHashing.swift
    1_stdlib/UnsafePointer.swift.gyb

- LLDB: both OSX and Ubuntu

build-script --debug-swift --debug-lldb --test -- --skip-test-cmark --skip-test-swift --lldb-use-system-debugserver

Timeout in TestREPLQuickLookObject.REPLQuickLookTestCase.testREPL

It won't be clear whether this is related until my revert passes, but it seems likely.

austinzheng commented:
Hi @atrick,

I can't build the project using the `--preset=buildbot` config presented above; it complains about not having some other variables set. I'm trying to reproduce with `./utils/build-script --swift-stdlib-assertions --swift-assertions --debug-swift --debug-swift-stdlib --test`; do you think this is sufficient?

atrick commented:
That should be sufficient, but if you can just define some dummy variables like this:

build-script <options> install_destdir=$install_dir installable_package=$build_name.tar.gz

austinzheng commented:
Hi @atrick,

I checked out the latest master, updated my dependencies, backed out your commit, and ran the tests on my machine using the following command:

`./utils/build-script --swift-stdlib-assertions --swift-assertions --debug-swift --debug-swift-stdlib --clean --test`.

However, the unit test suite appears to complete successfully for me. (OS X 10.11.1, MBP 15")

I'm going to try defining the dummy vars and running the preset.

atrick commented:
The linux test failure, Timeout in TestREPLQuickLookObject.REPLQuickLookTestCase.testREPL, went away after reverting this pull request.

It will take the bots a long time to verify that the Reflection tests are fixed. Meanwhile I'm also trying to reproduce them locally.

austinzheng commented:
@atrick Is this failure only on Linux? Do you know if it passed on OS X?

gribozavr commented:
I think the failure was on iphonesimulator-i386.  You can run tests with `ninja check-swift-iphonesimulator-i386` in the build directory.

austinzheng commented:
Excellent, I'm running the test suite now. I wonder if my "bigger than
Int64.max" test broke something on that architecture.

For future reference, what's the maximal set of tests I should run before
filing another PR? Should I run the check-swift command against all the
Apple architectures in the build directory?

Thanks,
Austin

On Wed, Jan 27, 2016 at 2:02 PM, Dmitri Gribenko <notifications@github.com>
wrote:

> I think the failure was on iphonesimulator-i386. You can run tests with ninja
> check-swift-iphonesimulator-i386 in the build directory.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift/pull/1058#issuecomment-175882513>.
>

gribozavr commented:
Usually `build-script -RT` is sufficient.  Sometimes a pull request can change some platform-dependent behavior, like i386 simulator, but it is hard to tell in advance.  You can run tests for extra platforms if you want to be extra safe, but we can't require that for practical reasons.

austinzheng commented:
Alright, that's super helpful.

I see the i386 breaking behavior; I can fix it when I get back tonight.
However, not sure about the timeout Linux issue; I'll have to dig into that
further.

One question for now: I thought UInt and Int were always 64-bit, as per the
stdlib documentation. However, it seems like they are variable width like
the equivalent C types, hence "error: integer overflows when converted from
'UInt64' to 'UInt'". Is that true, or documented anywhere?

Austin

On Wed, Jan 27, 2016 at 2:12 PM, Dmitri Gribenko <notifications@github.com>
wrote:

> Usually build-script -RT is sufficient. Sometimes a pull request can
> change some platform-dependent behavior, like i386 simulator, but it is
> hard to tell in advance. You can run tests for extra platforms if you want
> to be extra safe, but we can't require that for practical reasons.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift/pull/1058#issuecomment-175887171>.
>

gribozavr commented:
Thanks!

> However, it seems like they are variable width like the equivalent C types

Yes, that's correct -- Int and UInt are of the same size as the platform word (32 bits on i386 iphonesimulator).

atrick commented:
Sorry, the LLDB failure is not linux specific. The timeout on TestREPLQuickLookObject reproduces on OSX.

build-script --debug-swift --debug-lldb --test -- --skip-test-cmark --skip-test-swift --lldb-use-system-debugserver

=== BUILD TARGET repl_swift OF PROJECT lldb WITH CONFIGURATION CustomSwift-Debug ===

Timeout in TestREPLQuickLookObject.REPLQuickLookTestCase.testREPL

@scallanan I haven't reproduced this myself yet. If you have trouble, maybe someone from lldb can help.
